vantage/budget_resource_model.go

```diff
@@ -0,0 +1,237 @@
+		CostReportToken: src.CostReportToken.ValueString(),
+		WorkspaceToken:  src.WorkspaceToken.ValueString(),
+	}
+
+	if !src.Periods.IsNull() && !src.Periods.IsUnknown() {
+		periods := make([]*budgetPeriodResourceModel, 0, len(src.Periods.Elements()))
+		if diag := src.Periods.ElementsAs(ctx, &periods, false); diag.HasError() {
+			diags.Append(diag...)
+			return nil
+		}
+
+		dstValues := make([]*modelsv2.CreateBudgetPeriodsItems0, 0, len(periods))
+		for _, p := range periods {
+			periodItem := &modelsv2.CreateBudgetPeriodsItems0{
+				Amount: p.Amount.ValueFloat64Pointer(),
+			}
+
+			if p.EndAt.String() != "<unknown>" {
+				endAt, err := time.Parse("2006-01-02", strings.ReplaceAll(p.EndAt.String(), "\"", ""))
+				if err != nil {
+					diags.AddError("parsing error", fmt.Sprintf("failed to parse end_at: %s", err))
+					return nil
+				}
+
+				ea := strfmt.Date(endAt)
+				periodItem.EndAt = &ea
+			}
+
+			startAt, err := time.Parse("2006-01-02", strings.ReplaceAll(p.StartAt.String(), "\"", ""))
+			if err != nil {
+				diags.AddError("parsing error", fmt.Sprintf("failed to parse start_at: %s", err))
+				return nil
+			}
+
+			if !startAt.IsZero() {
+				sa := strfmt.Date(startAt)
+				periodItem.StartAt = &sa
+			}
+			dstValues = append(dstValues, periodItem)
+		}
+		dst.Periods = dstValues
+	}
+	return dst
+}
+
+func toUpdateModel(ctx context.Context, diags *diag.Diagnostics, src budgetModel) *modelsv2.UpdateBudget {
+	dst := &modelsv2.UpdateBudget{
+		Name:            src.Name.ValueString(),
+		CostReportToken: src.CostReportToken.ValueString(),
+	}
+
+	if !src.Periods.IsNull() && !src.Periods.IsUnknown() {
+		periods := make([]*budgetPeriodResourceModel, 0, len(src.Periods.Elements()))
+		if diag := src.Periods.ElementsAs(ctx, &periods, false); diag.HasError() {
+			diags.Append(diag...)
+			return nil
+		}
+
+		dstValues := make([]*modelsv2.UpdateBudgetPeriodsItems0, 0, len(periods))
+		for _, p := range periods {
+			periodItem := &modelsv2.UpdateBudgetPeriodsItems0{
+				Amount: p.Amount.ValueFloat64Pointer(),
+			}
+
+			if p.EndAt.String() != "<unknown>" {
+				endAt, err := time.Parse("2006-01-02", strings.ReplaceAll(p.EndAt.String(), "\"", ""))
+				if err != nil {
+					diags.AddError("parsing error", fmt.Sprintf("failed to parse end_at: %s", err))
+					return nil
+				}
+
+				ea := strfmt.Date(endAt)
+				periodItem.EndAt = &ea
+			}
+
+			startAt, err := time.Parse("2006-01-02", strings.ReplaceAll(p.StartAt.String(), "\"", ""))
+			if err != nil {
+				diags.AddError("parsing error", fmt.Sprintf("failed to parse start_at: %s", err))
+				return nil
+			}
+
+			if !startAt.IsZero() {
+				sa := strfmt.Date(startAt)
+				periodItem.StartAt = &sa
+			}
+			dstValues = append(dstValues, periodItem)
+		}
+		dst.Periods = dstValues
+	}
+
+	return dst
+}
+
+func applyBudgetPayload(ctx context.Context, isDataSource bool, src *modelsv2.Budget, dst *budgetModel) diag.Diagnostics {
```

Comment:
@conner n.b., unlike what you did, the model is not the receiver of this function, because i'm using the autogenerated model, (`budgetModel` is a ` resource_budget.BudgetModel`) and can't attach methods to it.